lib/jekyll/reader.rb

```diff
@@ -0,0 +1,113 @@
+# encoding: UTF-8
+require 'csv'
+
+module Jekyll
+  class Reader
+    attr_reader :site
+
+    def initialize(site)
+      @site = site
+    end
+
+    # Read Site data from disk and load it into internal data structures.
+    #
+    # Returns nothing.
+    def read
+      @site.layouts = LayoutReader.new(site).read
+      read_directories
+      @site.data = DataReader.new(site).read(site.config['data_source'])
+      CollectionReader.new(site).read
+    end
+
+    # Recursively traverse directories to find posts, pages and static files
+    # that will become part of the site according to the rules in
+    # filter_entries.
+    #
+    # dir - The String relative path of the directory to read. Default: ''.
+    #
+    # Returns nothing.
+    def read_directories(dir = '')
+      retrieve_posts(dir)
+
+      # Obtain sub-directories in order to recursively read them.
```

Comment:
Comments in methods indicate to me code that is need of further refactoring towards expressiveness. What about a more functional style approach to filtering out the directories?